### ISSUE TITLE
Why yolo inference results change when using different batch_size?

### ISSUE THREAD

**Issue by apacuk** [ultralytics/yolov5#11618]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

When I evaluate yolov5 model I get slightly different results (detection boxes that differ by up to few pixels) when I change `batch_size`. I've tested it both on default models and models I've fine-tuned. What is the reason?

Steps to reproduce:

1. install virtual environment (I've used `python=3.9`) as in `README.md`
2. Run following code:

```
import torch
import numpy as np
import random

model = torch.hub.load("ultralytics/yolov5", "yolov5s")
url = "https://ultralytics.com/images/zidane.jpg"

imgs_1 = [url]
print(model(imgs_1).pandas().xyxy[0])

imgs_2 = [url, url] # for simplicity we evaluate the same img twice
print(model(imgs_2).pandas().xyxy[0])
```

### Additional

I've tried setting random seeds and enforcing determinism (by prepending each model inference call with following code), but it didn't help. Problem exists both when evaluating on cpu only and when using gpu.

```
torch.manual_seed(0)
random.seed(0)
np.random.seed(0)
torch.use_deterministic_algorithms(True)
```

**Comment by apacuk** [comment]
According to this 
[discussion](https://discuss.pytorch.org/t/different-outputs-when-using-different-batch-size-only-on-cuda/155886/2 ), there is no guarantee that in this use case I would get the same results (for different `batch_size`). Kudos @a-zapala .

**Comment by glenn-jocher** [comment]
Hello @apacuk! Thank you for trying out YOLOv5 and expressing your concerns regarding the difference in results when using different batch sizes. You are right that changing the batch size can yield different results, not only in YOLOv5 but in neural networks in general. This is because neural networks are trained using stochastic optimization (stochastic gradient descent) which introduces randomness during training.

As mentioned in the discussion you linked, batch size can have an impact on the amount of randomness introduced during the inference process, causing slightly different results in some cases. However, the difference in results should not be significant and should not affect the overall accuracy of the model.

Thank you to @a-zapala for providing useful insight on the matter. If you have any further concerns or questions, please don't hesitate to let us know!
